### PR TITLE
Analysis._format_hook_datas() fixed for #1190.

### PR DESCRIPTION
**Pull request by leycec**
This method was fundamentally broken when passed a directory rather than glob or file. Specifically, directory walking accumulated invalid directories via the local `dest_dir` variable. This corrects that, thereby also correcting issue #1190 and hence matplotlib integration.

To support these changes, this method has been lightly refactored. Single-and double-letter variable names have been renamed for human-readability, multi-line comments and docstrings have been added, and various subtle errors in the original have been amended.

~~In particular, the original implementation assumed that existing paths which do *not* satisfy `os.path.isfile(pathname)` must be directories. This was an erroneous assumption: since `os.path.isfile(pathname)` returns False for *all* most special files (e.g., device nodes, sockets, and symlinks to such files), such implementation would have erroneously attempted to walk non-directory special files as if they were directories. (**Yeah.** That's not gonna work.)~~ **EDIT:** *Surprisingly, that does appear to work. Well, sort of. When passed a non-directory, `os.walk()` returns the empty list rather than raising an error. Ignore this paragraph. It is bad.*

This changeset has been tested on a reasonably complex Python 3 application importing `matplotlib`, `numpy`, and `scipy`. Happily, everything appears to work as advertised now.

### CODE REVIEW

File: PyInstaller/build.py [hunk 1]
```diff
@@ -368,26 +368,61 @@ def __init__(self, scripts=None, pathex=None, hiddenimports=None,
...
+        for src_root_path_or_glob, trg_root_dir in getattr(hook, 'datas', []):
+            for src_root_path in glob.glob(src_root_path_or_glob):
+                if os.path.isfile(src_root_path):
+                    toc_datas.append((
+                        os.path.join(
+                            trg_root_dir, os.path.basename(src_root_path)),
+                        src_root_path, 'DATA'))
+                elif os.path.isdir(src_root_path):
+                    # If no top-level target directory was passed, default this
+                    # to the basename of the top-level source directory.
```
Comment (htgoebel) [inline PyInstaller/build.py]: Shouldn't this better raise an error?

File: PyInstaller/build.py [hunk 2]
```diff
@@ -368,26 +368,61 @@ def __init__(self, scripts=None, pathex=None, hiddenimports=None,
...
+                    if not trg_root_dir:
+                        trg_root_dir = os.path.basename(src_root_path)
+
+                    for src_dir, src_subdir_basenames, src_file_basenames in\
+                        os.walk(src_root_path):
```
Comment (htgoebel) [inline PyInstaller/build.py]: Please at least put a space in front of the backslash if using line-continuation.

File: PyInstaller/build.py [hunk 3]
```diff
@@ -368,26 +368,61 @@ def __init__(self, scripts=None, pathex=None, hiddenimports=None,
...
+                        # of the passed top-level source directory. Since
+                        # os.walk() does *NOT* follow symlinks by default, this
+                        # should be the case. (But let's make sure.)
+                        assert src_dir.startswith(src_root_path)
+
+                        # Relative path of the current target directory,
+                        # obtained by removing the top-level source directory
+                        # from the current source directory (e.g., removing
+                        # "/top" from "/top/dir").
+                        trg_dir = trg_root_dir + src_dir[len(src_root_path):]
```
Comment (htgoebel) [inline PyInstaller/build.py]: Why not using `os.path.relpath()` here?

**Comment by htgoebel** [comment]
Thanks a lot! for this patch and the clean-up. Yes, this really has been an *ugly* part, using string-concatenation for path-manipulation.
@matysek For me this looks perfectly okay. This only change in semantic that I found is: Now only directories are walked recursivly. But since this is a quite central part, I'd want do have your opinion prior to merging.

File: PyInstaller/build.py (hunk 1 above)
Comment (leycec) [inline PyInstaller/build.py]: Sure. Maybe? The codebase currently assumes that `trg_root_dir` when empty will default to the basename of the top-level source directory. If we change that assumption to raise exceptions when `trg_root_dir` is empty, we'll need to significantly expand the patch.

Specifically, every hook currently setting such field to the empty string (e.g., `hook-babel.py`, `hook-matplotlib.py`) will need to be refactored to set such field to the expected basename. This is definitely feasible, but fairly extensive and hence error-prone. Perhaps we could hold off on that by prefixing `_format_hook_datas()` with a FIXME comment explaining the desired change?

File: PyInstaller/build.py (hunk 3 above)
Comment (leycec) [inline PyInstaller/build.py]: **Because you're smarter than me, Hartmut.**

...actually, I didn't know about `os.path.relpath()`. As you say, that's what we should be doing here. I'll make it so.

**Comment by leycec** [comment]
> Thanks a lot! for this patch and the clean-up.

No worries, guys. I suppose this was *my* way of saying: "Thank you." PyInstaller's awesome, and you're the reason why.

> This only change in semantic that I found is: Now only directories are walked recursively.

If my commit did change semantics, it didn't mean to and I'd better fix that. But *did* it change semantics? Both the current implementation and my commit should walk directories and symlinks to directories – and nothing else. (`os.walk()` appears to return the empty list when passed a non-directory or symlink to a non-directory.)

Or did I miss something? I probably missed something.

**Comment by htgoebel** [comment]
>> This only change in semantic that I found is: Now only directories are walked recursively.

> If my commit did change semantics, it didn't mean to and I'd better fix that.

No need for you to fix anything. Your changes are correct. As you wrote: the former implementation was plain wrong. (I just want do have a peer-review by @matysek prior to merging.)

File: PyInstaller/build.py (hunk 1 above)
Comment (htgoebel) [inline PyInstaller/build.py]: No, you are smarter than me :-)
You are right: This is a existing behaviour and we should not change it (yet). Adding a FIXME-comment is a good idea, though.

Comment (leycec) [inline PyInstaller/build.py]: > No, you are smarter than me :-)

Oh, *you*... :wink: 

> Adding a FIXME-comment is a good idea, though.

**Done.** It's a bit wordy, but that's probably better than "...uhm, what was we gonna do again?"

**Comment by leycec** [comment]
> No need for you to fix anything.

Those are my favorite words ever.

**So.** I've fixed up the prior commit with... a new commit! (Let me know if you'd like me to squash that.) Where, oh where art thou, gentle @matsyek?

**Comment by leycec** [comment]
I've fixed a subtle issue affecting both the current version and my revision of `_format_hook_datas()`. Unexpectedly, `glob.glob()` returns the empty list rather than raising an exception when passed non-existent paths: e.g.,

    >>> import glob
    >>> glob.glob('/this/should/raise/exceptions/but/it/doesnt')
    []

That's bad, implying that PyInstaller currently accepts non-existent paths in the `datas` list of module hooks *without* printing errors. This is trivial to demonstrate: e.g.,

    # Add this to any hook used by the current application. No errors will be printed.
    datas = [('/this/should/raise/exceptions/but/it/doesnt', '')]

This commit corrects that by explicitly testing for list emptiness and raising an exception instead. **Take that, Guido.**

**Comment by htgoebel** [comment]
@leycec https://github.com/pyinstaller/pyinstaller/pull/1194#issuecomment-80052825 is a good thing.

As I did not head any comment from @matysek, I'm merging this now.

Thanks for this valuable patch!

**Comment by leycec** [comment]
Thanks for being awesome, Hartmut. :fist:
